polling_stations/apps/data_finder/helpers.py

```diff
@@ -7,99 +7,222 @@
+    def geocode(self):
+        addresses = Address.objects.filter(postcode=self.postcode)
+        if not addresses:
+            raise ObjectDoesNotExist('No addresses found for postcode %s' % (self.postcode))
+
+        codes = self.get_codes(self.get_uprns(addresses))
+        centre = centre_from_points_qs(addresses)
+        return {
+            'source': 'addressbase',
+            'wgs84_lon': centre.x,
+            'wgs84_lat': centre.y,
+            'council_gss': codes['council_gss'],
+            'gss_codes': codes['gss_codes'],
+        }
 
-        try:
-            # attempt to parse error from json
-            res_json = res.json()
-            if 'error' in res_json:
-                raise PostcodeError("Mapit error {}: {}".format(res_json['code'], res_json['error']))
-            else:
-                raise PostcodeError("Mapit error {}: unknown".format(res.status_code))
-        except ValueError:
-            # if we fail to parse json, raise a less specific exception
-            raise PostcodeError("Mapit error {}: unknown".format(res.status_code))
+    def geocode_point_only(self):
+        addresses = Address.objects.filter(postcode=self.postcode)
+        if not addresses:
+            raise ObjectDoesNotExist('No addresses found for postcode %s' % (self.postcode))
 
-    res_json = res.json()
+        centre = centre_from_points_qs(addresses)
+        return {
+            'source': 'addressbase',
+            'wgs84_lon': centre.x,
+            'wgs84_lat': centre.y,
+        }
+
+
+def geocode_point_only(postcode, sleep=True, mapit=mapit_wrapper_factory):
+    addressbase = AddressBaseWrapper(postcode)
+    mapit = mapit(postcode)
+
+    try:
+        # first try addressbase
+        result = addressbase.geocode_point_only()
+    except ObjectDoesNotExist:
+        # we couldn't find this postcode in AddressBase: fall back to mapit
+
+        # optional sleep to avoid hammering mapit
+        if sleep:
+            time.sleep(1.3)
+
+        result = mapit.geocode()
+    except:
+        # something else went wrong: lets give mapit a go anyway
+
+        # optional sleep to avoid hammering mapit
+        if sleep:
+            time.sleep(1.3)
 
-    gss_codes = []
-    council_gss = None
-    for area in res_json['areas']:
-        if 'gss' in res_json['areas'][area]['codes']:
-            gss = res_json['areas'][area]['codes']['gss']
-            gss_codes.append(gss)
-            if res_json['areas'][area]['type'] in COUNCIL_TYPES:
-                council_gss = gss
-
-    if 'wgs84_lon' not in res_json or 'wgs84_lat' not in res_json:
-        raise PostcodeError("No location information")
-
-    return {
-        'wgs84_lon': res_json['wgs84_lon'],
-        'wgs84_lat': res_json['wgs84_lat'],
-        'gss_codes': gss_codes,
-        'council_gss': council_gss,
-    }
+        result = mapit.geocode()
+
+    return result
+
+
+def geocode(postcode, mapit=mapit_wrapper_factory):
```

Comment:
> I'm not sure what `mapit_wrapper_factory` is doing

The reason for `mapit_factory` is for tesing/mocking purposes - its just a way to make `MapitWrapper` an injected dependency.

By passing in a function which return an instance of `MapitWrapper` as a default parameter, when we are writing tests we can easily create a mock `MapitWrapper` class and pass in a `mock_factory` instead. This means we can more easily abstract the behaviour of `MapitWrapper` when testing the exception hadling in `geocode()` (where we don't care about the actual response data - just which class returned it) without having to use really verbose VCR fixtures. See https://github.com/DemocracyClub/UK-Polling-Stations/pull/651/files#diff-1ef176bdf330b9e534a5489f367dac51R5

I'd really like to keep the ability to do that rather than have to use VCR to fake the HTTP responses in this instance. It may be that there is a more 'pythonic' way of injecting this than using a factory though. Suggestions?

> or why `AddressBaseWrapper` doesn't have one

The reason I didn't do this for `AddressBaseWrapper` is because there are no HTTP calls being made to mock. I defined the behaviour for `AddressBaseWrapper` in those tests by just re-using the existing addressbase DB fixture that was needed for the tests in https://github.com/DemocracyClub/UK-Polling-Stations/pull/651/files#diff-98027f40280a91d59bb6094c405daeeaR8 anyway.


> I think the approach here should be to have a list of classes that can do the geocoding, in the order we want

Yes. I do like your idea in points 2 and 3 about making the geocoding sources more generic/flexible so that we can swap sources in and out more easily. I will think through the details of that a bit more this afternoon.